src/CodeGen_Hexagon.cpp

```diff
@@ -0,0 +1,1383 @@
+#include <iostream>
+#include <sstream>
+
+#include "LLVM_Headers.h"
+#include "CodeGen_Hexagon.h"
+#include "IROperator.h"
+#include "IRMatch.h"
+#include "IREquality.h"
+#include "Target.h"
+#include "Debug.h"
+#include "Util.h"
+#include "Simplify.h"
+#include "IRPrinter.h"
+#include "EliminateBoolVectors.h"
+#include "HexagonOptimize.h"
+#include "AlignLoads.h"
+#include "CSE.h"
+#include "LoopCarry.h"
+
+#ifdef WITH_HEXAGON
+
+#define IPICK(i64) (B128 ? i64##_128B : i64)
+
+namespace Halide {
+namespace Internal {
+
+using std::vector;
+using std::string;
+using std::ostringstream;
+using std::pair;
+using std::make_pair;
+
+using namespace llvm;
+
+CodeGen_Hexagon::CodeGen_Hexagon(Target t) : CodeGen_Posix(t) {}
+
+std::unique_ptr<llvm::Module> CodeGen_Hexagon::compile(const Module &module) {
+    auto llvm_module = CodeGen_Posix::compile(module);
+    static bool options_processed = false;
+
+    // TODO: This should be set on the module itself, or some other
+    // safer way to pass this through to the target specific lowering
+    // passes. We set the option here (after the base class'
+    // implementaiton of compile) because it is the last
+    // Hexagon-specific code to run prior to invoking the target
+    // specific lowering in LLVM, minimizing the chances of the wrong
+    // flag being set for the wrong module.
+    if (!options_processed) {
+      cl::ParseEnvironmentOptions("halide-hvx-be", "HALIDE_LLVM_ARGS",
+                                  "Halide HVX internal compiler\n");
+      // We need to EnableQuIC for LLVM and Halide (Unrolling).
+      char *s = strdup("HALIDE_LLVM_QUIC=-hexagon-small-data-threshold=0");
```

Comment:
Cool, doing this now. What does this option actually do? The comment mentions unrolling, I'd like to be a bit more precise...